apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerCommand.cs

```diff
@@ -0,0 +1,367 @@
+        /// <returns></returns>
+        public Task<T> ExecuteScalarAsync<T>()
+        {
+            return ExecuteScalarAsync<T>(CancellationToken.None);
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <typeparam name="T"></typeparam>
+        /// <returns></returns>
+        public async Task<T> ExecuteScalarAsync<T>(CancellationToken cancellationToken)
+        {
+            using (var reader = await ExecuteDbDataReaderAsync(CommandBehavior.SingleRow, cancellationToken)
+                .ConfigureAwait(false))
+            {
+                if (await reader.ReadAsync(cancellationToken).ConfigureAwait(false) && reader.HasRows &&
+                    reader.FieldCount > 0)
+                    return reader.GetFieldValue<T>(0);
+            }
+            return default(T);
+        }
+
+        /// <inheritdoc />
+        public override async Task<object> ExecuteScalarAsync(CancellationToken cancellationToken)
+        {
+            using (var reader = await ExecuteDbDataReaderAsync(CommandBehavior.SingleRow, cancellationToken)
+                .ConfigureAwait(false))
+            {
+                if (await reader.ReadAsync(cancellationToken).ConfigureAwait(false) && reader.HasRows &&
+                    reader.FieldCount > 0)
+                    return reader[0];
+            }
+            return null;
+        }
+
+        /// <inheritdoc />
+        public override void Prepare()
+        {
+            //Spanner does not support preoptimized queries nor 2 phase commit transactions.
+        }
+
+        /// <inheritdoc />
+        protected override DbParameter CreateDbParameter()
+        {
+            return new SpannerParameter();
+        }
+
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+        }
+
+        /// <inheritdoc />
+        protected override DbDataReader ExecuteDbDataReader(CommandBehavior behavior)
+        {
+            ValidateCommandBehavior(behavior);
+            return ExecuteDbDataReaderAsync(behavior, _synchronousCancellationTokenSource.Token).Result;
+        }
+
+        /// <inheritdoc />
+        protected override async Task<DbDataReader> ExecuteDbDataReaderAsync(CommandBehavior behavior,
+            CancellationToken cancellationToken)
+        {
+            // There must be a valid and open connection.
+            if (SpannerConnection == null)
+                throw new InvalidOperationException(
+                    "You must assign a SpannerConnection to this command to execute it.");
+            if (SpannerCommandTextBuilder.SpannerCommandType != SpannerCommandType.Select)
+                throw new InvalidOperationException("You can only call ExecuteReader on a Select Command");
+            if (!SpannerConnection.IsOpen)
+                await SpannerConnection.OpenAsync(cancellationToken).ConfigureAwait(false);
+            if (!SpannerConnection.IsOpen)
```

Comment:
the extra exception is probably unnecessary, and is just extra protection.
calling open simultaneously either explicitly or implicitly via running multiple commands is not supported and OpenAsync will throw an error blocking it.
You *can* run multiple commands with a single connection as long as you open it beforehand.
concurrent read operations will share the single reserved spanner session.

if you execute a write operation it will also use the single reserved spanner session if no other read operation is currently running.  This is so you can serially execute several mixed read/writes using only the single reserved session.  It's only when you start concurrently running reads with writes or multiple concurrent writes that additional sessions will be acquired from the global session pool.

So TLDR is -- as long as you open the connection up front, you can actually do anything you want in terms of concurrent operations.  
What SpannerConnection provides is automatically reserving a session up front (reserved during the open operation which historically has had bad perf in other providers) and sharing that session across your commands until your concurrent operations force additional sessions to be needed.